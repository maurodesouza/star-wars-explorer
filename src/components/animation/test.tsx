import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Animation, AnimationProps } from '.';

jest.mock('react-lottie', () => ({
  __esModule: true,
  default: function Mock() {
    return <svg data-testid="lottie" />;
  },
}));

const dataMock: AnimationProps = {
  animation: 'loading',
  label: 'foo',
};

describe('<Animation />', () => {
  it('should render the animation styles correct', () => {
    const { container } = renderWithProviders(<Animation {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the animation component correct', () => {
    renderWithProviders(<Animation {...dataMock} />);

    const label = screen.getByText(dataMock.label!);
    const lottie = screen.getByTestId('lottie');

    expect(label).toBeInTheDocument();
    expect(lottie).toBeInTheDocument();
  });
});
