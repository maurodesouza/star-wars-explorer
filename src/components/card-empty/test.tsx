import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { EmptyCard } from '.';

const dataMock = {
  id: 'test',
};

describe('<EmptyCard />', () => {
  it('should render the empty card styles correct', () => {
    const { container } = renderWithProviders(<EmptyCard {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the empty card component correct', () => {
    renderWithProviders(<EmptyCard {...dataMock} />);

    const label = screen.getByText(`${dataMock.id}#`);

    expect(label).toBeInTheDocument();
  });
});
