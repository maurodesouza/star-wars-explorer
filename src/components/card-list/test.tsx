import { cleanup } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';
import { createArray } from 'utils';

import { CardList } from '.';

jest.mock('components', () => ({
  __esModule: true,
  Card: function Mock() {
    return <div data-testid="card" />;
  },
}));

const itemMock = {
  id: '1',
  title: 'C-3PO',
  image: 'image',
  entity: Entities.CHARACTERS,
};

describe('<CardList />', () => {
  describe('styles', () => {
    beforeAll(() => {
      jest.resetAllMocks();
    });

    it('should render the card list styles correct', () => {
      const { container } = renderWithProviders(<CardList />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('should render the items card correct', () => {
    const counts = [4, 8];

    counts.forEach(count => {
      const items = createArray(count).map((_, index) => ({
        ...itemMock,
        id: String(index),
      }));

      const render = renderWithProviders(<CardList items={items} />);
      const findedItems = render.getAllByTestId(/card/i);

      expect(findedItems.length).toBe(count);

      cleanup();
    });
  });

  it('should render the fills correct', () => {
    const minCountToFill = 4;
    const counts = [1, 3, 4, 6];

    counts.forEach(count => {
      const result = count - minCountToFill;
      const lackForFill = result <= 0 ? Math.abs(result) : 0;

      const items = createArray(count).map((_, index) => ({
        ...itemMock,
        id: String(index),
      }));

      const { container } = renderWithProviders(<CardList items={items} />);
      const findedItems = container.firstElementChild!.querySelectorAll(
        'div:not([data-testId]'
      );

      expect(findedItems.length).toBe(lackForFill);
      cleanup();
    });
  });
});
