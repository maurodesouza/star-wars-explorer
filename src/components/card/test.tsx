/* eslint-disable @typescript-eslint/no-var-requires */

import { act, fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';
import { Card } from '.';

const useRouter = jest.spyOn(require('next/router'), 'useRouter');
const push = jest.fn();

useRouter.mockImplementation(() => ({
  push,
}));

const mockData = {
  id: '1',
  label: 'C-3PO',
  image:
    'https://raw.githubusercontent.com/tbone849/star-wars-guide/master/build/assets/img/characters/2.jpg',
};

describe('<Card />', () => {
  it('should render the card correct', () => {
    renderWithProviders(<Card {...mockData} />);

    expect(
      screen.getByRole('img', { name: mockData.label })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: mockData.label })
    ).toBeInTheDocument();
  });

  it("should render the no image found element when can't load image url", async () => {
    renderWithProviders(<Card {...mockData} />);

    const image = screen.getByRole('img', { name: mockData.label });

    act(() => {
      fireEvent.error(image);
    });

    const noImageFoundElement = screen.getByText(/no found/i);

    expect(noImageFoundElement).toBeInTheDocument();
  });

  it('should render the card styles correct', () => {
    const { container } = renderWithProviders(<Card {...mockData} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should navigate to entity page', () => {
    const { container } = renderWithProviders(<Card {...mockData} />);

    fireEvent.click(container.firstChild!);

    expect(push).toBeCalledTimes(1);
    expect(push).toBeCalledWith(`/${Entities.CHARACTERS}/${mockData.label}`);
  });
});
