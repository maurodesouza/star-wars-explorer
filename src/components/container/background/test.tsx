import { renderWithProviders } from 'utils/test/helpers';
import { Background } from '.';

describe('components/container', () => {
  describe('<Background />', () => {
    it('should render the background styles correct', () => {
      const { container } = renderWithProviders(<Background />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
