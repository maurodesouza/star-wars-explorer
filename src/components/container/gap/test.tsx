import { renderWithProviders } from 'utils/test/helpers';
import { Gap } from '.';

describe('components/container', () => {
  describe('<Gap />', () => {
    it('should render the gap styles correct', () => {
      const { container } = renderWithProviders(<Gap />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
