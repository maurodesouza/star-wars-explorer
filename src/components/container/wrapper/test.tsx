import { renderWithProviders } from 'utils/test/helpers';
import { Wrapper } from '.';

describe('components/container', () => {
  describe('<Wrapper />', () => {
    it('should render the wrapper styles correct', () => {
      const { container } = renderWithProviders(<Wrapper />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
