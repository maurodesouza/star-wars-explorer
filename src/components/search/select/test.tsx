import { act, fireEvent, screen } from '@testing-library/react';

import { renderWithProviders } from 'utils/test/helpers';

import { events } from 'app';
import { Entities } from 'types';

import { Select } from '.';

jest.useFakeTimers();

describe('<Select />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  it('should render the select styles correct', () => {
    const { container } = renderWithProviders(<Select />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should open dropdown select on input click and hide on blur', () => {
    const { container } = renderWithProviders(<Select />);

    const input = screen.getByLabelText(/open drop down select/i);

    const dropdown = container.children[0].children[2];

    expect(dropdown.getAttribute('aria-hidden')).toBe('true');

    act(() => {
      fireEvent.click(input);
    });

    expect(dropdown.getAttribute('aria-hidden')).toBe('false');

    act(() => {
      fireEvent.blur(input);
    });

    expect(dropdown.getAttribute('aria-hidden')).toBe('true');
  });

  it('should close dropdown after select some option', () => {
    const { container } = renderWithProviders(<Select />);

    const input = screen.getByLabelText(/open drop down select/i);

    const dropdown = container.children[0].children[2];

    act(() => {
      fireEvent.click(input);
    });

    act(() => {
      fireEvent.click(dropdown.children[0]);
    });

    expect(dropdown.getAttribute('aria-hidden')).toBe('true');
  });

  it('should send search make event on select some option', async () => {
    const searchMakeSpy = jest.spyOn(events.search, 'make');
    renderWithProviders(<Select />);

    const input = screen.getByLabelText(/open drop down select/i);

    act(() => {
      fireEvent.click(input);
    });

    const option = screen.getByText(Entities.PLANETS);

    act(() => {
      fireEvent.click(option);
    });

    expect(searchMakeSpy).toBeCalledTimes(1);
    expect(searchMakeSpy).toBeCalledWith({ entity: Entities.PLANETS });
  });
});
