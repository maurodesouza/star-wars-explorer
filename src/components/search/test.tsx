import { act, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithProviders } from 'utils/test/helpers';

import { events } from 'app';
import { sleep } from 'utils';

import { Search, SearchProps } from '.';

jest.mock('./select', () => ({
  __esModule: true,
  Select: function Mock() {
    return <div data-testid="search select" />;
  },
}));

const searchProps: Required<SearchProps> = {
  label: 'label',
  placeholder: 'placeholder',
};

describe('<Search />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the search correct', () => {
    renderWithProviders(<Search {...searchProps} />);

    const input = screen.getByPlaceholderText(searchProps.placeholder);
    const label = screen.getByLabelText(searchProps.label);
    const select = screen.getByTestId(/search select/i);

    expect(input).toBeInTheDocument();
    expect(select).toBeInTheDocument();
    expect(label).toBeInTheDocument();
  });

  it('should render the search styles correct', () => {
    const { container } = renderWithProviders(<Search />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render error', () => {
    const errorText = 'some error';

    renderWithProviders(<Search />);

    act(() => {
      events.search.error(errorText);
    });

    expect(screen.getByText(errorText)).toBeInTheDocument();
  });

  it('should send search event with corret values after type', async () => {
    const searchMakeEventFn = jest.spyOn(events.search, 'make');
    const text = 'some value';

    renderWithProviders(<Search {...searchProps} />);

    const input = screen.getByPlaceholderText(searchProps.placeholder);

    userEvent.type(input, text);
    await sleep(400);

    expect(searchMakeEventFn).toBeCalledWith({ search: text });
    expect(searchMakeEventFn).toBeCalledTimes(1);
  });

  it('should send search event with corret values after submit form with Enter', async () => {
    const searchMakeEventFn = jest.spyOn(events.search, 'make');

    renderWithProviders(<Search {...searchProps} />);

    const form = screen.getByRole('form');

    act(() => {
      fireEvent.submit(form);
    });

    expect(searchMakeEventFn).toBeCalledWith({ search: '' });
    expect(searchMakeEventFn).toBeCalledTimes(1);
  });
});
