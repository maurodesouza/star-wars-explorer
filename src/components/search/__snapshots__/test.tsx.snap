// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render the search styles correct 1`] = `
.c0 {
  width: min(100%,40rem);
  position: relative;
  margin-top: 3.2rem;
  margin-bottom: 4.0rem;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 1.6rem;
  color: #FFF;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-top: 0.8rem;
}

.c3 {
  width: 100%;
  background-color: #FFF;
  padding: 1.2rem;
  color: #000;
  font-size: 1.6rem;
  border: 0;
  border-radius: 0 0.8rem 0.8rem 0;
}

.c3::-webkit-input-placeholder {
  color: #000;
}

.c3::-moz-placeholder {
  color: #000;
}

.c3:-ms-input-placeholder {
  color: #000;
}

.c3::placeholder {
  color: #000;
}

@media (min-width:768px) {
  .c0 {
    width: min(100%,54rem);
    margin-top: calc(4.0rem * 1.5);
    margin-bottom: 4.8rem;
  }
}

@media (min-width:768px) {
  .c1 {
    font-size: 2.4rem;
  }
}

@media (min-width:768px) {
  .c2 {
    margin-top: 1.2rem;
  }
}

@media (min-width:768px) {
  .c3 {
    padding: 2.4rem;
    font-size: 2.4rem;
  }
}

<form
  aria-label="form"
  class="c0"
>
  <label
    class="c1"
    for="search"
  >
    Explore by
    <div
      class="c2"
    >
      <div
        data-testid="search select"
      />
      <input
        class="c3"
        id="search"
        placeholder="Ex: Luke"
      />
    </div>
  </label>
</form>
`;
